Tests/ApolloCacheDependentTests/ReadWriteFromStoreTests.swift

```diff
@@ -0,0 +1,265 @@
+    ]
+    
+    try withCache(initialRecords: initialRecords) { (cache) in
+      let store = ApolloStore(cache: cache)
+      
+      try await(store.withinReadTransaction { transaction in
+        XCTAssertThrowsError(try transaction.readFragment(ofType: HeroDetails.self, withKey: "2001")) { error in
+          if case let error as GraphQLResultError = error {
+            XCTAssertEqual(error.path, ["primaryFunction"])
+            XCTAssertMatch(error.underlying, JSONDecodingError.missingValue)
+          } else {
+            XCTFail("Unexpected error: \(error)")
+          }
+        }
+      })
+    }
+  }
+  
+  func testReadFriendsNamesFragment() throws {
+    let initialRecords: RecordSet = [
+      "QUERY_ROOT": ["hero": Reference(key: "2001")],
+      "2001": [
+        "name": "R2-D2",
+        "__typename": "Droid",
+        "friends": [
+          Reference(key: "1000"),
+          Reference(key: "1002"),
+          Reference(key: "1003")
+        ]
+      ],
+      "1000": ["__typename": "Human", "name": "Luke Skywalker"],
+      "1002": ["__typename": "Human", "name": "Han Solo"],
+      "1003": ["__typename": "Human", "name": "Leia Organa"],
+      ]
+
+    try withCache(initialRecords: initialRecords) { (cache) in
+      let store = ApolloStore(cache: cache)
+
+      try await(store.withinReadTransaction { transaction in
+        let friendsNamesFragment = try transaction.readFragment(ofType: FriendsNames.self, withKey: "2001")
+
+        let friendsNames = friendsNamesFragment.friends?.flatMap { $0?.name }
+        XCTAssertEqual(friendsNames, ["Luke Skywalker", "Han Solo", "Leia Organa"])
+      })
+    }
+  }
+  
+  func testUpdateFriendsNamesFragment() throws {
+    let initialRecords: RecordSet = [
+      "QUERY_ROOT": ["hero": Reference(key: "2001")],
+      "2001": [
+        "name": "R2-D2",
+        "__typename": "Droid",
+        "friends": [
+          Reference(key: "1000"),
+          Reference(key: "1002"),
+          Reference(key: "1003")
+        ]
+      ],
+      "1000": ["__typename": "Human", "name": "Luke Skywalker"],
+      "1002": ["__typename": "Human", "name": "Han Solo"],
+      "1003": ["__typename": "Human", "name": "Leia Organa"],
+    ]
+
+    try withCache(initialRecords: initialRecords) { (cache) in
+      let store = ApolloStore(cache: cache)
+
+      try await(store.withinReadWriteTransaction { transaction in
+        try transaction.updateFragment(ofType: FriendsNames.self, withKey: "2001") { (friendsNames: inout FriendsNames) in
+          friendsNames.friends?.append(.init(__typename: "Droid", name: "C-3PO"))
+        }
+      })
+
+      let result = try await(store.load(query: HeroAndFriendsNamesQuery()))
+      guard let data = result.data else { XCTFail(); return }
+
+      XCTAssertEqual(data.hero?.name, "R2-D2")
+      let friendsNames = data.hero?.friends?.flatMap { $0?.name }
+      XCTAssertEqual(friendsNames, ["Luke Skywalker", "Han Solo", "Leia Organa", "C-3PO"])
+    }
+  }
+}
```

Comment:
Yeah, I definitely need to add tests with variables, and make them pass :) Because there are some remaining issues with that.